src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,394 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
+ * accessing ownerThread unless we know the current thread already owns the monitor.
+ * Only the owner of the DebugRawMonitor will ever modify ownerThread. The main thing
+ * we are trying to protect from is a thread checking to see if it is the owner of
+ * the DebugRawMonitor, and having the ownerThread field change while doing this.
+ * The check is made using:
+ *    isSameObject(env, dbg_monitor->ownerThread, current_thread)
+ * We don't want ownerThread changing in the middle of this check since it can lead
+ * to a crash if the JNI handle gets freed. So this check is always protected by
+ * dbgRawMonitor as are modifications to ownerThread.
+ *
+ * There is also one other minor role that dbgRawMonitor plays. In verifyMonitorRank()
+ * we iterate over all monitors, and need to make sure the monitor we are checking has
+ * been initialized. If it hasn't been, then it is skipped. There can be a race and
+ * we catch it in the middle of intialization. Since dbgRawMonitor is held during
+ * verifyMonitorRank(), we should also hold it during monitor initialization.
+ */
+
+static jrawMonitorID dbgRawMonitor;
+
+static void
+createDbgRawMonitor()
+{
+    JDI_ASSERT(gdata->rankedMonitors);
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,CreateRawMonitor)
+            (gdata->jvmti, "Debug Raw Monitor", &dbgRawMonitor);
+    if (error != JVMTI_ERROR_NONE) {
+        EXIT_ERROR(error, "on creation of dbgRawMonitor");
+    }
+    JDI_ASSERT(dbgRawMonitor != NULL);
+}
+
+void
+dbgRawMonitor_lock()
+{
+    JDI_ASSERT(gdata->rankedMonitors);
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,RawMonitorEnter)
+            (gdata->jvmti, dbgRawMonitor);
+    error = ignore_vm_death(error);
+    if (error != JVMTI_ERROR_NONE) {
+      EXIT_ERROR(error, "on dbgRawMonitor enter");
+    }
+}
+
+void
+dbgRawMonitor_unlock()
+{
+    JDI_ASSERT(gdata->rankedMonitors);
+    jvmtiError error;
+    error = JVMTI_FUNC_PTR(gdata->jvmti,RawMonitorExit)
+            (gdata->jvmti, dbgRawMonitor);
+    error = ignore_vm_death(error);
+    if (error != JVMTI_ERROR_NONE) {
+      EXIT_ERROR(error, "on dbgRawMonitor exit");
+    }
+}
+
+DebugRawMonitor*
+debugMonitorCreate(DebugRawMonitorRank dbg_monitor_rank, char *name)
 {
     jrawMonitorID monitor;
     jvmtiError error;
+    DebugRawMonitor* dbg_monitor = &dbg_monitors[dbg_monitor_rank];
+    JDI_ASSERT(dbg_monitor->monitor == NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,CreateRawMonitor)
                 (gdata->jvmti, name, &monitor);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on creation of a raw monitor");
     }
-    return monitor;
+
+    JDI_ASSERT(monitor != NULL);
+
+    // Need to lock during initialization so verifyMonitorRank() can be guaranteed that
+    // if the monitor field is set, then the monitor if fully initialized. More specifically,
+    // that the rank field has been set.
```

Comment:
rank for all monitors can be set in `util_initialize()`:
```
for (i = 0; i < NUM_DEBUG_RAW_MONITORS; i++) {
  dbg_monitors[i]->rank = i;
}
```

Reply:
Ok.

Reply:
I decided not to make this change. It was a bit disruptive with the need for forward declarations. Plus I wasn't convinced I wouldn't still need the locking since there is also the assignment of the name field (although that is only used if something has gone wrong) and also I was a little worried that tearing could happen on some platforms (a partial ptr value is read in), although not on any platform that we support. I did update the comment to make it less specific about the rank field.